// @remix-run/node/dist/magicExports/remix.js
/**
 * @remix-run/node v1.19.3
 *
 * Copyright (c) Remix Software Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.md file in the root directory of this source tree.
 *
 * @license MIT
 */
'use strict';
Object.defineProperty(exports, '__esModule', { value: true });
var node = require('@remix-run/node');
/** @deprecated Import `createCookie` from `@remix-run/node` instead. */
exports.createCookie = node.createCookie;

/** @deprecated Import `createCookieSessionStorage` from `@remix-run/node` instead. */
exports.createCookieSessionStorage = node.createCookieSessionStorage;

/** @deprecated Import `createFileSessionStorage` from `@remix-run/node` instead. */
exports.createFileSessionStorage = node.createFileSessionStorage;

/** @deprecated Import `createMemorySessionStorage` from `@remix-run/node` instead. */
exports.createMemorySessionStorage = node.createMemorySessionStorage;

/** @deprecated Import `createSessionStorage` from `@remix-run/node` instead. */
exports.createSessionStorage = node.createSessionStorage;

/** @deprecated Import `unstable_createFileUploadHandler` from `@remix-run/node` instead. */
exports.unstable_createFileUploadHandler = node.unstable_createFileUploadHandler;

/** @deprecated Import `unstable_createMemoryUploadHandler` from `@remix-run/node` instead. */
exports.unstable_createMemoryUploadHandler = node.unstable_createMemoryUploadHandler;

/** @deprecated Import `unstable_parseMultipartFormData` from `@remix-run/node` instead. */
exports.unstable_parseMultipartFormData = node.unstable_parseMultipartFormData;
